File: RileyLink/RFPacket.m
//
//  RFPacket.m
//  RileyLink
//
//  Created by Pete Schwamb on 2/28/16.
//  Copyright © 2016 Pete Schwamb. All rights reserved.
//

#import "RFPacket.h"

@implementation RFPacket

- (instancetype)init NS_UNAVAILABLE
{
  return nil;
}

- (instancetype)initWithData:(NSData*)data
{
  self = [super init];
  if (self) {
    if (data.length > 0) {
      unsigned char rssiDec = ((const unsigned char*)[data bytes])[0];
      unsigned char rssiOffset = 73;
      if (rssiDec >= 128) {
        self.rssi = (short)((short)( rssiDec - 256) / 2) - rssiOffset;
      } else {
        self.rssi = (rssiDec / 2) - rssiOffset;
      }
    }
    if (data.length > 1) {
      self.packetNumber = ((const unsigned char*)[data bytes])[1];
    }
    
    if (data.length > 2) {
      _data = [data subdataWithRange:NSMakeRange(2, data.length - 2)];
    }
  }
  return self;
}

@end

File: RileyLinkBLEKit/RFPacket.swift
//
//  RFPacket.swift
//  RileyLinkBLEKit
//
//  Created by Pete Schwamb on 9/16/17.
//  Copyright © 2017 Pete Schwamb. All rights reserved.
//

import Foundation

public struct RFPacket {
    public let data: Data
    let packetCounter: Int
    public let rssi: Int

    public init?(rfspyResponse: Data) {
        guard rfspyResponse.count > 2 else {
            return nil
        }
        
        let rssiDec = Int(rfspyResponse[0])
        let rssiOffset = 73
        if rssiDec >= 128 {
            self.rssi = (rssiDec - 256) / 2 - rssiOffset
        } else {
            self.rssi = rssiDec / 2 - rssiOffset
        }

        self.packetCounter = Int(rfspyResponse[1])
        
        self.data = rfspyResponse.subdata(in: 2..<rfspyResponse.count)
    }
}

